### PR TITLE
txn: Let commit rollback pessimistic lock instead of committing as WriteType::Lock

### LINKED ISSUE #14551: Stale pessimistic lock requests in force-lock mode may lead to data corruption or in consistency after resolving lock
## Bug Report

Allowing locking with conflict introduces more corner cases if we consider the posibility of stale RPC requests.

To resolve stale pessimistic locks (mostly caused by pessimistic-rollback failures) of committed transaction, we commit them as `WriteType::Lock` (introduced in https://github.com/tikv/tikv/pull/6353):

https://github.com/tikv/tikv/blob/68298d834be1844eaf254d5237eed4856605833c/src/storage/txn/actions/commit.rs#L44-L59

But it seems we don't need to commit it since the key is no longer part of that transaction. I can't remember the reason of committing them as Lock records. Probably it's because it's easier to handle in the `commit` function (doesn't need to introduce rollback operation to the `commit` procedure).

Now, as locking with conflict is sometimes allowed, we have such a new corner case:

1. Transaction T starts (start_ts = 10)
2. T locks key k1 with force-locking is enabled, then the key is locked at for_update_ts = 10. Then the lock is `pessimisticRollback`-ed.
3. Transaction T commits (start_ts = 10, commit_ts = 20). It finally doesn't involve key k1.
4. Transaction T2 writes key k1 and committed (start_ts = 15, commit_ts = 20).
5. A stale pessimistic lock request in step-2 arrives at TiKV, and successfully locked k1 anyway (since locking-with-conflict is enabled).
6. Another transaction T3 encounters the lock written in step-5, and starts resovle locking. It then find that transaction T is successfully committed, so it calls `commit` on the stale pessimistic lock.
7. The logic of committing pessimistic lock mentioned above is executed, so a `Lock` record is written to write cf with commit_ts = 20. However, T2 has a commit record at commit_ts = 20 on this key too, so it's overwritten by the `Lock` record of transaction T.

In such case, transaction T2 may lost its change to key k1. If k1 is T2's primary, it may also affect the state of T2's secondaries.

A way to fix it is to rollback the pessimistic lock instead of committing as WriteType::Lock when trying to commit a pessimistic lock.

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Close #14551

What's Changed:

```commit-message
Changes the behavior of `commit` meeting pessimistic lock, from committing as WriteType::Lock to rolling-back. It's correct considering that the key is nolonger part of that transaction. This change fixes the problem that stale pessimistic lock requests with force-locking enabled (which is used by TiDB in fair-locking mode) may overwrite the commit record of another transaction and cause data loss.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Fix an issue that in some rare cases, stale pessimistic lock requests sent by TiDB with `tidb_pessimistic_txn_fair_locking` enabled may affect data correctness.
```

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@cfzjywxk @ekexium PTAL when you have time, and also please help confirm if this solution is correct, thanks!

File: src/storage/txn/actions/commit.rs [hunk 1]
```diff
@@ -368,4 +380,29 @@ pub mod tests {
             assert_eq!(write.txn_source, source);
         }
     }
+
+    #[test]
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/storage/txn/actions/commit.rs:384]: The test can be added to TiDB or client-go. But I'm not sure if it's possible to construct same commit_ts for two transactions from the client side. I'll have a try later.

File: src/storage/txn/actions/commit.rs [hunk 2]
```diff
@@ -45,18 +45,19 @@ pub fn commit<S: Snapshot>(
             // our transaction model. But a pessimistic lock will be left if the pessimistic
             // rollback request fails to send and the transaction need not to acquire this
```
Comment (ekexium) [inline src/storage/txn/actions/commit.rs:46]: Another possible reason is duplicated stale `AcquireLock` request, even without `PessimisticRollback`?

**Review by ekexium — APPROVED** [review]
> It's correct considering that the key is nolonger part of that transaction.

More accurately I think the condition is that the pessimistic lock does not have to be committed. Possibilities:
(1) It should be pessimistically rolled back, i.e. not part of the txn. Removing the lock is equivalent to that the `PessimisticRollback` arrives earlier.
(2) The original transaction contains it.
(2.1) If the txn rolled back, we roll back the lock as well.
(2.2) If the txn is in progress, a loss of a pessimistic lock will not affect correctness.
(2.3) If the txn is committed, it must be a stale `AcquireLock` request. We can remove the lock.

**Comment by MyonKeminta** [comment]
> (2.1) If the txn rolled back, we roll back the lock as well.
> (2.2) If the txn is in progress, a loss of a pessimistic lock will not affect correctness.

For a resolve_locks request, if `commit` is called, it must infer that it has confirmed that the transaction is in a committed state.

However if the commit-secondaries phase (or commit phase of async commit transactions), here might be such a problem as stated in this issue:
* https://github.com/pingcap/tidb/issues/42937

I didn't find anything else abnormal apart from that.

Another thing that I'm afraid of is that whether there's any code that may rely on the fact that TiKV can commit a pessimistic lock as a `WriteType::Lock` record, so that it commits without prewriting intentionally.

**Comment by MyonKeminta** [comment]
If we want to be safer, we can do the rollback only when lock.for_update_ts >= commit_ts, in which case we can say that the pessimistic lock definitely shouldn't exist. The commit-as-lock-record logic will be still kept as is, even if it's indeed unnecessary...

**Comment by MyonKeminta** [comment]
How do you think? @cfzjywxk @ekexium

**Comment by cfzjywxk** [comment]
I think just rollback the pessimistic lock if safer, rollback pessimistic locks does not affect transaction status but commit them may.

**Review by cfzjywxk — APPROVED** [review]

**Comment by cfzjywxk** [comment]
@ekexium PTAL again

**Review by ekexium — APPROVED** [review]

**Pull request merged**